### PR TITLE
Create-Add files to greedy_method directory

### PR DESCRIPTION
**Pull request by Apoorve73**
## Greedy Method approach is one of the method to solve Optimisation Problems and is widely used. 
The file **knapsack_problem.py** is an example problem to show the working of **greedy approach,** along with a **knapsack_unittest** file.

* [x] Add an algorithm?
* [ ] Fix a bug or typo in an existing algorithm?
* [ ] Documentation change?

### **Checklist:**
* [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [x] This pull request is all my own work -- I have not plagiarized.
* [x] I know that pull requests will not be merged if they fail the automated tests.
* [x] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [ ] All new Python files are placed inside an existing directory.
* [x] All filenames are in all lowercase characters with no spaces or dashes.
* [x] All functions and variable names follow Python naming conventions.
* [x] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [x] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Comment by cclauss** [comment]
How is this faster or better than https://github.com/TheAlgorithms/Python/search?q=knapsack ?

https://docs.pytest.org/en/latest/goodpractices.html#conventions-for-python-test-discovery
says that pytest will run unit tests in `test_*.py` or `*_test.py` files so please modify the test filename.

**Comment by cclauss** [comment]
I am not a fan of creating a new directory here.  What do other maintainers think given the [other implementations](https://github.com/TheAlgorithms/Python/search?q=knapsack).

**Comment by Apoorve73** [comment]
> How is this faster or better than https://github.com/TheAlgorithms/Python/search?q=knapsack ?
> 
> https://docs.pytest.org/en/latest/goodpractices.html#conventions-for-python-test-discovery
> says that pytest will run unit tests in `test_*.py` or `*_test.py` files so please modify the test filename.

Its faster in terms of Time Complexiity. Greedy methods are generally faster. For example, Dijkstra's shortest path algorithm takes O(ELogV + VLogV) time. Dynamic Programming is generally slower. For example, Bellman Ford algorithm takes O(VE) time.

I will change the filename. Thank you @cclauss  :+1:

**Comment by Apoorve73** [comment]
> I am not a fan of creating a new directory here. What do other maintainers think given the [other implementations](https://github.com/TheAlgorithms/Python/search?q=knapsack).

Well! Greedy Algo is not the same as Dynamic Programming, so I created a different directory for the same. You may tell me, if these files need to be shifted in an existing directory. :100:

**Comment by cclauss** [comment]
Can we please have a benchmark that ___proves___ that this implementation is faster than all other knapsack implementations in this repo.

**Comment by Apoorve73** [comment]
> Can we please have a benchmark that _**proves**_ that this implementation is faster than all other knapsack implementations in this repo.

Yeah sure! Greedy Algorithm is not the best but in some cases, it is! Refer https://www.geeksforgeeks.org/greedy-approach-vs-dynamic-programming/ for a detailed explanation! :)

File: greedy_method/test_*.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import unittest
```
Comment (cclauss) [inline greedy_method/test_*.py:1]: Please change the filename to `test_knapsack.py`

Comment (Apoorve73) [reply greedy_method/test_*.py:1]: Sure thing! Thank you @cclauss !

File: greedy_method/knapsack_problem.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        # encountered before.
+        if 15 - limit >= weight[index]:
+
+            limit += weight[index]
+            # Adding profit gained for the given weight 1 ===
+            # weight[index]/weight[index]
+            gain += 1 * profit[index]
+
+        else:
+            # Since the weight encountered is greater than limit, therefore take the required number of remaining kgs
```
Comment (cclauss) [inline greedy_method/knapsack_problem.py:75]: Please wrap lines at 88 characters max.

File: greedy_method/knapsack_problem.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    max_Weight = int(input())
+
+    if len(profit) != len(weight):
+        print("The length of both the arrays must be same! Try again!")
+
+    if max_Weight < 0:
+        print("Gotcha! Weight is a positive quantity")
+
+    for i in range(len(profit)):
+        if (profit[i] or weight[i]) < 0:
```
Comment (cclauss) [inline greedy_method/knapsack_problem.py:100]: Can we please have two tests where profit[i] is < zero and and where weight[i] is < zero because I do not think that this code works the way that you think it does.

File: greedy_method/test_knapsack.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+    """
+
+    def test_sorted(self):
+        """
+        kp.calc_Profit takes the required argument (profit, weight, max_weight)
+        and returns whether the answer matches to the expected ones
+        """
+        gain = kp.calc_Profit([10, 20, 30, 40, 50, 60], [2, 4, 6, 8, 10, 12], 100)
+        # self.assertTrue(result)
+        self.assertEqual(gain, 75)
```
Comment (cclauss) [inline greedy_method/test_knapsack.py:17]:
```suggestion
        assert gain == 75
```
Pytest does not need all the boilerplate code.

Can we please also have tests that does negative integers, zeros, and floats point numbers so we can test negative user situations?

File: greedy_method/knapsack_problem.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    # List created to store profit gained for the 1kg in case of each weight
+    # respectively
+    profit_By_Weight = list()
+
+    # Calculate and append profit/weight for each
+    for i in range(len(weight)):
+        profit_By_Weight.append(profit[i] / weight[i])
+
+    # Creating a copy of the list and sorting proit/weight in ascending order
+    temp_PBW = sorted(copy.deepcopy(profit_By_Weight))
```
Comment (cclauss) [inline greedy_method/knapsack_problem.py:42]: I doubt that `copy.deepcopy()` is needed here.

File: greedy_method/knapsack_problem.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+
+    # declaring useful variables
+    length = len(temp_PBW)
+    limit = 0
+    gain = 0
+    i = 0
+
+    # loop till the total weight do not reach max limit i.e. 15 kg and till i<l
+    while limit <= 15 and i < length:
```
Comment (cclauss) [inline greedy_method/knapsack_problem.py:51]:
```suggestion
    while i < limit <= 15:
```

File: greedy_method/knapsack_problem.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    return gain
+
+
+if __name__ == "__main__":
+
+    # Input profit, weight and max_Weight (all positive values).
+
+    profit = [int(x) for x in input().split()]
+    weight = [int(x) for x in input().split()]
+    max_Weight = int(input())
```
Comment (cclauss) [inline greedy_method/knapsack_problem.py:91]:
```suggestion
    max_Weight = int(input("<Let's put some useful user prompt here...>"))
```

File: greedy_method/knapsack_problem.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+
+    if len(profit) != len(weight):
+        print("The length of both the arrays must be same! Try again!")
```
Comment (cclauss) [inline greedy_method/knapsack_problem.py:94]:
```suggestion
        raise ValueError("The length of both the arrays must be same! Try again!")
```
This would be more Pythonic.

**Comment by Apoorve73** [comment]
Thank you @cclauss for all the valuable suggestions. I will fix them soon!

File: greedy_method/knapsack_problem.py (hunk 5 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:42]: Yeah! Since profit_By_Weight list is not directly assigned to temp_PBW, there is no need of any shallow or deepcopy operations. I have fixed it! Thanks :)

File: greedy_method/knapsack_problem.py (hunk 6 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:51]: It may be ```i <= limit <= 15 ``` since initially both i and limit is 0.

File: greedy_method/knapsack_problem.py (hunk 2 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:75]: Solved!

File: greedy_method/knapsack_problem.py (hunk 7 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:91]: Yeah! Fixed it ! Thanks @cclauss

File: greedy_method/knapsack_problem.py (hunk 8 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:94]: That's much better than simple print statement! Fixed! :+1:

File: greedy_method/knapsack_problem.py (hunk 3 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:100]: IT works fine, but that's true, that it doesn't work exactly the way I want it to! Fixed

File: greedy_method/test_knapsack.py (hunk 4 above)
Comment (Apoorve73) [reply greedy_method/test_knapsack.py:17]: Sure thing! Will add them too!

**Comment by Apoorve73** [comment]
@cclauss kindly review the changes!

File: greedy_method/knapsack_problem.py [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+Constraints:
+Max_Weight > 0
+profit[i] >= 0
+weight[i] >= 0
+Calculate the max profit the shopkeeper can make for the given max_Weight that could be carried.
+"""
+from typing import Union
+
+
+def Calc_Profit(profit: list, weight: list, max_Weight: int) -> Union[str, int]:
```
Comment (mateuszz0000) [inline greedy_method/knapsack_problem.py:21]: Function names should be lower_case.

Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:21]: Okay! Thank you :)

File: greedy_method/knapsack_problem.py [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+                "<< Oops! Could not accept a negative value for weight. Try Again.. >>"
+            )
+
+    # List created to store profit gained for the 1kg in case of each weight
+    # respectively
+    profit_By_Weight = list()
+
+    # Calculate and append profit/weight for each
+    for i in range(len(weight)):
+        profit_By_Weight.append(profit[i] / weight[i])
```
Comment (mateuszz0000) [inline greedy_method/knapsack_problem.py:61]:
```
profit_by_weight = [p / w for p, w in zip(profit, weight)]
```

File: greedy_method/knapsack_problem.py (hunk 9 above)
Comment (mateuszz0000) [reply greedy_method/knapsack_problem.py]: variables, function params too :smiley:

File: greedy_method/knapsack_problem.py (hunk 10 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py:61]: Fixed :+1:

File: greedy_method/knapsack_problem.py (hunk 9 above)
Comment (Apoorve73) [reply greedy_method/knapsack_problem.py]: Sure thing :)

**Comment by Apoorve73** [comment]
Thank you @cclauss  and @mateuszz0000 for assisting me. I have tried my best to fix all the issues and improve code quality. Kindly review it, so as to get my PR merged or to suggest any further changes.

**Comment by Apoorve73** [comment]
@cclauss , is there any more changes to be performed? :)

**Comment by Apoorve73** [comment]
Thanks once again! :+1:
